compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -282,12 +282,29 @@ public void visit(BLangLiteral literalExpr) {
         }
 
         // Check whether this belongs to decimal type or float type
-        if (literalType.tag == TypeTags.FLOAT) {
-            if (expType.tag == TypeTags.DECIMAL) {
+        if (literalType.tag == TypeTags.FLOAT || literalType.tag == TypeTags.DECIMAL) {
+            if (expType.tag == TypeTags.FLOAT && literalType.tag == TypeTags.DECIMAL) {
+                dlog.error(literalExpr.pos, DiagnosticCode.INCOMPATIBLE_TYPES, expType, literalType);
+            }
+
+            String numericLiteral = String.valueOf(literalValue);
+            char lastChar = getLastChar(numericLiteral);
+
+            if (expType.tag == TypeTags.DECIMAL && (lastChar == 'f' || lastChar == 'F')) {
+                dlog.error(literalExpr.pos, DiagnosticCode.INCOMPATIBLE_TYPES, literalType, expType);
+            }
+
+            if (expType.tag == TypeTags.DECIMAL && !(lastChar == 'f' || lastChar == 'F')) {
                 literalType = symTable.decimalType;
-                literalExpr.value = String.valueOf(literalValue);
-            } else if (expType.tag == TypeTags.FLOAT) {
-                literalExpr.value = Double.parseDouble(String.valueOf(literalValue));
+            }
+
+            numericLiteral = stripFloatDecimalDescriminater(numericLiteral, lastChar);
+
+            if (literalType.tag == TypeTags.DECIMAL) {
+                literalType = symTable.decimalType;
+                literalExpr.value = numericLiteral;
```

Comment:
Can we include this line within the check at L297? With that, it seems we can remove condition at L303.